crates/engine/tree/src/tree/payload_processor/prewarm.rs

```diff
@@ -78,37 +86,72 @@ where
     }
 
     /// Spawns all pending transactions as blocking tasks by first chunking them.
+    ///
+    /// For Optimism chains, special handling is applied to the first transaction if it's a
+    /// deposit transaction (type 0x7E/126) which sets critical metadata that affects all
+    /// subsequent transactions in the block.
     fn spawn_all(
         &self,
-        pending: mpsc::Receiver<impl ExecutableTxFor<Evm> + Send + 'static>,
+        pending: mpsc::Receiver<impl ExecutableTxFor<Evm> + Clone + Send + 'static>,
         actions_tx: Sender<PrewarmTaskEvent>,
     ) {
         let executor = self.executor.clone();
         let ctx = self.ctx.clone();
         let max_concurrency = self.max_concurrency;
+        let transaction_count_hint = self.transaction_count_hint;
 
         self.executor.spawn_blocking(move || {
-            let mut handles = Vec::with_capacity(max_concurrency);
             let (done_tx, done_rx) = mpsc::channel();
             let mut executing = 0;
-            while let Ok(executable) = pending.recv() {
-                let task_idx = executing % max_concurrency;
 
-                if handles.len() <= task_idx {
-                    let (tx, rx) = mpsc::channel();
-                    let sender = actions_tx.clone();
-                    let ctx = ctx.clone();
-                    let done_tx = done_tx.clone();
+            // Initially cap workers based on transaction count hint
+            let mut handles = Vec::with_capacity(max_concurrency);
+            let workers_needed = transaction_count_hint.min(max_concurrency);
 
-                    executor.spawn_blocking(move || {
-                        ctx.transact_batch(rx, sender, done_tx);
-                    });
+            // Spawn the required number of workers
+            for _ in 0..workers_needed {
+                handles.push(ctx.spawn_worker(&executor, actions_tx.clone(), done_tx.clone()));
+            }
 
-                    handles.push(tx);
+            // Handle first transaction - special case for Optimism deposit / systems txs
+            if let Ok(first_tx) = pending.recv() {
+                if first_tx.tx().is_type(OPTIMISM_DEPOSIT_TX_TYPE) {
```

Comment:
we can do > 4 here because this will catch all non eth types and if its first position we can assume this is a sys tx

Reply:
this makes sense, thanks for the note!

  - Type 0: Legacy transactions (original Ethereum transactions)
  - Type 1: EIP-2930 transactions (with access lists)
  - Type 2: EIP-1559 transactions (with base fee + priority fee)
  - Type 3: EIP-4844 transactions (blob transactions for L2 data)
  - Type 4: EIP 7702